torsiondrive/qm_engine.py

```diff
@@ -327,6 +335,166 @@ def load_native_output(self, filename='output.dat'):
+        if self.native_opt:
+            assert self.temp_type == 'optimize', "input_file should be a opt job to use native opt"
+        self.gauss_temp = gauss_temp
+        self.M = Molecule()
+        self.M.elem = elems
+        self.M.xyzs = [np.array(coords, dtype=float)]
+        self.M.build_topology()
+
+    def write_input(self, filename='gaussian.com'):
+        """ Write Gaussian input using Molecule Class """
+        assert hasattr(self, 'gauss_temp'), "self.gauss_temp not set, call load_input() first"
+        with open(filename, 'w') as outfile:
+            for line in self.gauss_temp:
+                if line == '$!geometry@here':
+                    for e, c in zip(self.M.elem, self.M.xyzs[0]):
+                        outfile.write("%-7s %13.7f %13.7f %13.7f\n" % (e, c[0], c[1], c[2]))
+
+                elif line == "$!optblock@here":
+                    if hasattr(self, 'optblockStr'):
+                        # self.optblockStr will be set by self.optimize_native()
+                        outfile.write(self.optblockStr)
+
+                else:
+                    outfile.write(line)
+
+    def optimize_native(self):
+        """
+        Run the constrained optimization, following Gaussian09 manual.
+        1. write a optimization job input file.
+        2. run the job
+        """
+        assert self.temp_type == 'optimize', "To use native optimization, the input file be an opt job"
+        if self.extra_constraints is not None:
+            raise RuntimeError('extra constraints not supported in Gaussian09 native optimizations')
+        self.optblockStr =''
+        for d1, d2, d3, d4, v in self.dihedral_idx_values:
+            self.optblockStr += f'{d1 + 1} {d2 + 1} {d3 + 1} {d4 + 1} ={v:.3f} B\n'  # Build the angle
+            self.optblockStr += f'{d1 + 1} {d2 + 1} {d3 + 1} {d4 + 1} F\n'           # Freeze the angle
+        # write input file
+        self.write_input('gaussian.com')
+        # run the job
+        self.run('g09 < gaussian.com > gaussian.log', input_files=['gaussian.com'], output_files=['gaussian.log'])
+        self.run('formchk lig.chk lig.fchk')    # More reliable to get the geometry from the log file and convert it
+
+    def load_native_output(self, filename='lig.fchk', filename2='gaussian.log'):
+        """ Load the optimized geometry and energy into a new molecule object and return """
+        found_opt_result = False
+        final_energy, elems, coords = None, [], []
+        with open(filename2) as logfile:
+            logfile = logfile.readlines()
+            for counter, line in enumerate(logfile):
+                line = line.strip()
+                if line.startswith('Optimization completed'):
+                    found_opt_result = True
+
+        if found_opt_result is not True:
+            raise RuntimeError("Geometry optimisation failed in %s" % filename2)
+
+        with open(filename) as outfile:
+            outfile = outfile.readlines()
```

Comment:
This is not needed

Reply:
This is also the check that the geometry optimization has completed successfully.